### PR TITLE
Very rough approach for implementing language settings

### PR DESCRIPTION
**Pull request by rowanwins**
Related to #3707 where it's suggested to changing the naming of some buttons are parts of the interface.

This approach would see languageOverrides to be passed in from TerriaMap to via the options to `Terria.js`

The main downside I've seen so far is that some text components occasionally include React components like Icons, for example [empty workbench has a message](https://github.com/TerriaJS/terriajs/blob/master/lib/ReactViews/SidePanel/SidePanel.jsx#L162
) which includes a lightbulb icon. The current approach doesn't render those react components. Someone much better with React would probably have a few ideas how to handle this.

What I've done in `TerriaMap/index.js` is 
````
import languageOverrides from './lib/Language/overrides';

terriaOptions.languageOverrides = languageOverrides;
````
These overrides then get merged into the defaults.

Anyway this is my very naive approach :)

CustomLanguage

### CODE REVIEW

File: lib/ReactViews/SidePanel/SidePanel.jsx [hunk 1]
```diff
@@ -150,19 +150,11 @@ const SidePanel = createReactClass({
...
-                  <li>Browse the Data Catalogue</li>
-                  <li>Load your own data onto the map</li>
-                </ul>
-                <p>
-                  <Icon glyph={Icon.GLYPHS.bulb} />
-                  <strong>TIP:</strong>{" "}
-                  <em>All your active data sets will be listed here</em>
-                </p>
+                <div
+                  dangerouslySetInnerHTML={{
```
Comment (soyarsauce) [inline lib/ReactViews/SidePanel/SidePanel.jsx]: Even though the defaults are loaded in by terria, we should probably use something like https://github.com/cure53/DOMPurify to sanitise this anyway before we forget (e.g. later down the track it gets loaded in from elsewhere)

File: lib/Language/defaults.js [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+      <div>Your workbench is empty</div>
+      <p>
+        <strong>Click &apos;Add data&apos; above to:</strong>
+      </p>
+      <ul>
+        <li>Browse the Data Catalogue</li>
+        <li>Load your own data onto the map</li>
+      </ul>
+      <p>
+        <Icon glyph={Icon.GLYPHS.bulb} />
```
Comment (soyarsauce) [inline lib/Language/defaults.js]: This works???! 😮

Comment (soyarsauce) [reply lib/Language/defaults.js]: Answering my own question: probably not, based off the bulb missing in the screenshot 😆

Comment (soyarsauce) [reply lib/Language/defaults.js]: Answering even more, I just read the actual PR description. Nevermind me?

**Comment by soyarsauce** [comment]
So there's a few ways around the icon problem, e.g. feeding in an actual react component as part of the "i18n"/language config, or the simplest way for now is to keep the existing components as "defaults" (with icon), then overrides means no icon-ability atm.

File: lib/Language/defaults.js (hunk 2 above)
Comment (rowanwins) [reply lib/Language/defaults.js]: You make me laugh @soyarsauce!

**Comment by rowanwins** [comment]
I'll do a quick audit tomorrow of how many places we use components within regular HTML, if it's only the lightbulb icon in that empty workbench effected then I'm happy to have no icon-ability :)

File: lib/ReactViews/SidePanel/SidePanel.jsx (hunk 1 above)
Comment (rowanwins) [reply lib/ReactViews/SidePanel/SidePanel.jsx]: So I'm wondering how we differentiate between those things which are just text strings vs those which we'll accept a html string for...
Or perhaps as a first cut we should only do strings?

**Comment by kring** [comment]
@rowanwins @soyarsauce why not use https://github.com/TerriaJS/terriajs/blob/master/lib/ReactViews/Custom/parseCustomMarkdownToReact.js to render the localized "strings". Then you can support icons by defining a custom component (https://github.com/TerriaJS/terriajs/blob/master/lib/ReactViews/Custom/registerCustomComponentTypes.js). It sanitizes the input too.

**Comment by rowanwins** [comment]
Good idea @kring - I'll give it a whirlybird!

**Comment by rowanwins** [comment]
Okay @kring & @soyarsauce - the latest commit adds support for Icon's as customComponents.

~~The other component that comes to mind that might be nice to support via custom text is the Button component, eg I'm thinking about the possibility of buttons in that empty workbench area...~~

~~Before I add support for the Buttons do you want to make sure you're happy with that handling for the Icon component?~~
EDIT: We don't use a fancy button component, just regular old buttons with a few classes work fine!

And are we happy to proceed and get this into `master`?

On a sidenote there is also some similar stuff that we do via the config.json in TerriaMap (eg we have a `feedbackPreamble` option - who knew!! At some stage it would be good to rationalise.

**Comment by rowanwins** [comment]
Theoretically the `Icon.jsx` also takes a `style` object as a prop however this would be a smidge fiddily to support as we'd basically be taking a string and converting it to a object. I reckon we just support styling via the `className` prop.

**Comment by AnaBelgun** [comment]
reminder to @soyarsauce to review this one

**Comment by soyarsauce** [comment]
This is awesome work @rowanwins & looks good, hope you don't mind I just extended so we can also use any component/jsx, if the latest commit looks good to you, let's merge it
https://github.com/TerriaJS/terriajs/pull/3742/commits/a7305b0005277d0fdea4ce9be2cb3f113338c0a9

File: lib/Models/Terria.js [hunk 3]
```diff
@@ -35,6 +35,9 @@ var Services = require("./Services");
 var TimeSeriesStack = require("./TimeSeriesStack");
 var ViewerMode = require("./ViewerMode");
 var findIndex = require("../Core/findIndex.js");
+
+var language = require("../Language/defaults.jsx").language;
```
Comment (soyarsauce) [inline lib/Models/Terria.js:39]: I just switched this back to cjs to be consistent with the rest of th efile

**Comment by rowanwins** [comment]
Looks good to me @soyarsauce - it doesn't seem like I can merge it because I can't review it because I opened the PR, but I've looked over it and tested locally and it all makes sense. 
I'll open a corresponding PR over in the TerriaMap repo.

**Review by soyarsauce — APPROVED** [review]

**Comment by soyarsauce** [comment]
Sorry @rowanwins I thought I approved w/ my single comment review

**Pull request merged**